src/transforms/log_to_metric.rs

```diff
@@ -271,55 +275,150 @@ fn render_template(template: &Template, event: &Event) -> Result<String, Transfo
+                                event,
+                                name,
+                                template,
+                                &mut result,
+                                &mut static_tags,
+                                &mut dynamic_tags,
+                            )?;
                         }
                     }
                 }
             }
+            for (k, v) in static_tags {
+                if let Some(discarded_v) = dynamic_tags.insert(k.clone(), v.clone()) {
+                    warn!(
+                        "Static static overrides dynamic tag. \
+                key: {}, value: {:?}, discarded value: {:?}",
+                        k, v, discarded_v
+                    );
+                };
+            }
             result.as_option()
         }
     })
 }
 
 fn render_tag_into(
     event: &Event,
-    name: &str,
-    template: &Option<Template>,
+    key_template: &Template,
+    value_template: &Option<Template>,
     result: &mut MetricTags,
+    static_tags: &mut HashMap<String, TagValue>,
+    dynamic_tags: &mut HashMap<String, TagValue>,
 ) -> Result<(), TransformError> {
-    let value = match template {
-        None => TagValue::Bare,
+    let key_s = match render_template(key_template, event) {
+        Ok(key_s) => key_s,
+        Err(TransformError::TemplateRenderingError(err)) => {
+            emit!(crate::internal_events::TemplateRenderingError {
+                error: err,
+                drop_event: false,
+                field: Some(key_template.get_ref()),
+            });
+            return Ok(());
+        }
+        Err(err) => return Err(err),
+    };
+    match value_template {
+        None => {
+            result.insert(key_s, TagValue::Bare);
+        }
         Some(template) => match render_template(template, event) {
-            Ok(result) => TagValue::Value(result),
-            Err(TransformError::TemplateRenderingError(error)) => {
+            Ok(value_s) => {
+                tag_expansion(key_s.clone(), value_s, result, static_tags, dynamic_tags);
+            }
+            Err(TransformError::TemplateRenderingError(value_error)) => {
                 emit!(crate::internal_events::TemplateRenderingError {
-                    error,
+                    error: value_error,
                     drop_event: false,
-                    field: Some(name),
+                    field: Some(template.get_ref()),
                 });
                 return Ok(());
             }
-            Err(other) => return Err(other),
+            Err(err) => return Err(err),
         },
     };
-    result.insert(name.to_string(), value);
     Ok(())
 }
 
+static RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^0-9A-Za-z_]").unwrap());
+fn slugify_text(input: String) -> String {
+    let result = RE.replace_all(&input, "_");
+    result.to_lowercase()
+}
+
+// borrow from src/sinks/loki/sink.rs
+fn tag_expansion(
```

Comment:
I didn't look deeply yet but it seems like we can benefit from introducing a util, maybe under `src/common` to avoid code duplication.

Reply:
It's a good idea! I could try it.